### PR TITLE
KEP-2258: Node log query documentation

### PR DESCRIPTION
**Pull request by aravindhp**
Add node log query documentation:
- Code: https://github.com/kubernetes/kubernetes/pull/96120
- Issue:  https://github.com/kubernetes/enhancements/issues/2258

### CODE REVIEW

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 1]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
 In Kubernetes clusters created by the `kube-up.sh` script, log rotation is configured by the `logrotate` tool.
 The `logrotate` tool rotates logs daily, or once the log size is greater than 100MB.
 
+## Log query
+
+{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
+
+To help with debugging issues on nodes, Kubernetes v1.27 introduces a feature that allows viewing logs of services
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:238]:
```suggestion
To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of system services
```

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 2]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+running on the node. To use the feature, ensure that the `NodeLogQuery`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) kubelet options, `EnableSystemLogHandler`
+and `EnableSystemLogQuery`, are enabled. The feature is available only to cluster administrators.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:241]:
```suggestion
running on the node. To use the feature, ensure that the `NodeLogQuery`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled for
that node, and that the kubelet configuration options `enableSystemLogHandler`  and
`enableSystemLogQuery` are both set to true.

You can try out this alpha feature on all your nodes, or on just a subset.
```

A node level administrator can try this out.

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 3]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+
+You can now query services on the node:
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:243]:
```suggestion
Provided you're authorized to do so, you can now query logs for system services on that node:
```

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 4]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+```shell
+kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=<insert-service-name-here>"
+```
+
+You can also query files in the `/var/log/` directory:
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:248]:
```suggestion
You can also fetch files, provided that the files are in a directory that the kubelet allows for log
fetches. For example, you can fetch a log from `/var/log` on a Linux node:
```

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 5]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+the operating system's native logger like journald or to a log file in `/var/log/`. The heuristics firsts checks the
+native logger and if that is not available attempts to retrieve the first logs from `/var/log/<servicename>` or
+`/var/log/<servicename>.log` or `/var/log/<servicename>/<servicename>.log`.
+
+The complete list of options that can be used are:
+
+Option | Description
+------ | -----------
+`boot` | boot show messages from a specific system boot
+`pattern` | pattern filters log entries by the provided regex pattern
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:263]:
```suggestion
`pattern` | pattern filters log entries by the provided regular expression
```

Ideally, specify the flavor of regexp that is supported.

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 6]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+`query` | query specifies services(s) or files from which to return logs (required)
+`sinceTime` | sinceTime is an RFC3339 timestamp from which to show logs
+`tailLines` | tailLines is used to retrieve the specified number of lines from the end of the log
+`untilTime`, untilTime is an RFC3339 timestamp until which to show logs
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:267]: Is this right?
```suggestion
`sinceTime` |  an [RFC3339](https://www.rfc-editor.org/rfc/rfc3339) timestamp from which to show logs (inclusive)
`untilTime` | an [RFC3339](https://www.rfc-editor.org/rfc/rfc3339) timestamp until which to show logs (inclusive)
`tailLines` | specify how many lines from the end of the log to retrieve; the default is to fetch the whole log
```

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 7]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+```shell
+kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=<insert-service-name-here>"
+```
+
+You can also query files in the `/var/log/` directory:
+```shell
+kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=/<insert-log-file-name-here>"
+```
+
+The query options also uses heuristics to retrieve logs if the the user is not aware if a given service is logging to
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:253]:
```suggestion
The kubelet uses heuristics to retrieve logs. This helps if you are not aware whether a given system service is
writing logs to
```

File: content/en/docs/concepts/cluster-administration/system-logs.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:1]: It's good to document an assumption we make: on Linux, we assume that service logs are available either via journald, or by reading files within `/var/log`.

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 8]
```diff
@@ -231,6 +231,47 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
+
+To help with debugging issues on nodes, Kubernetes v1.27 introduces a feature that allows viewing logs of services
+running on the node. To use the feature, ensure that the `NodeLogQuery`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) kubelet options, `EnableSystemLogHandler`
+and `EnableSystemLogQuery`, are enabled. The feature is available only to cluster administrators.
+
+You can now query services on the node:
+```shell
+kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=<insert-service-name-here>"
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-logs.md:245]:
```suggestion
# Fetch kubelet logs from a node named node-1.example
kubectl get --raw "/api/v1/nodes/node-1.example/proxy/logs/?query=kubelet"
```

I think the kubelet is a good example of a service we can expect to find on nodes.

**Review by sftim — COMMENTED** [review]
Thanks

Some early feedback; I hope this helps.

**Review by aravindhp — COMMENTED** [review]
Thanks for the review, @sftim. I have incorporated your feedback. The KEP also introduces new kubelet configuration option. Does the API doc generation for that happen separately or should I be generating that in this PR?

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 9]
```diff
@@ -231,6 +231,53 @@ Similar to the container logs, you should rotate system component logs in the `/
...
 
+## Log query
+
+{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
+
+To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of services
+running on the node. To use the feature, ensure that the `NodeLogQuery`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled for that node, and that the
+kubelet configuration options `enableSystemLogHandler` and `enableSystemLogQuery` are both set to true. On Linux
+we assume that service logs are available either via journald. On Windows we assume that service logs are available
```
Comment (tengqm) [inline content/en/docs/concepts/cluster-administration/system-logs.md:242]: "either" ??

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 10]
```diff
@@ -231,6 +231,53 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+in the application log provider. On both operating systems, logs are also available by reading files within
+`/var/log/`.
+
+A node level administrator can try out this alpha feature on all their nodes, or on just a subset. Provided you're
+authorized to do so, you can now query logs for system services on that node:
```
Comment (tengqm) [inline content/en/docs/concepts/cluster-administration/system-logs.md:247]: This message is unclear to me.
"node level administrator" means nothing in the context of Kubernetes.
The authorization we can enforce is about the access to the Kubernetes API.
It sounds like we are having a feature enabled by one group of people (node-level admins),
but used by another group of people (authorized using RBAC, with no dedicated built in roles).
Am I understanding this correctly?

File: content/en/docs/concepts/cluster-administration/system-logs.md [hunk 11]
```diff
@@ -231,6 +231,53 @@ Similar to the container logs, you should rotate system component logs in the `/
...
+```shell
+# Fetch kubelet logs from a node named node-1.example
+kubectl get --raw "/api/v1/nodes/node-1.example/proxy/logs/?query=kubelet"
+```
+
+You can also fetch files, provided that the files are in a directory that the kubelet allows for log
+fetches. For example, you can fetch a log from `/var/log` on a Linux node:
+```shell
+kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=/<insert-log-file-name-here>"
+```
```
Comment (tengqm) [inline content/en/docs/concepts/cluster-administration/system-logs.md:257]: There must be some implicit restrictions to the `/<insert-log-file-name-here>`.
It sounds like I'm able to read the whole file system if the "node admin"
has not taken some special care of permissions.
By default `kubelet` runs as a root ... that is so scary to me.

**Comment by mickeyboxell** [comment]
@aravindhp could you respond to @tengqm's comments when you have a moment? Tomorrow is the **Docs complete — All PRs reviewed and ready to merge** deadline and we would love to see this merged by then.

File: content/en/docs/concepts/cluster-administration/system-logs.md (hunk 9 above)
Comment (aravindhp) [reply content/en/docs/concepts/cluster-administration/system-logs.md]: Removed

File: content/en/docs/concepts/cluster-administration/system-logs.md (hunk 10 above)
Comment (aravindhp) [reply content/en/docs/concepts/cluster-administration/system-logs.md]: I was just going by [this](https://github.com/kubernetes/website/pull/40003#discussion_r1137249798) suggestion. I have however updated it to clarify that we only enforce access to node objects.

File: content/en/docs/concepts/cluster-administration/system-logs.md (hunk 11 above)
Comment (aravindhp) [reply content/en/docs/concepts/cluster-administration/system-logs.md:257]: The capability to view the `/var/log` directory using the logs endpoint via the proxy was already present in the kubelet. All this feature is doing is extending it to also query the journal logs. As part of that we also want to retain the existing functionality to view the log files in `/var/log` using the `--query` option. 

I also want to call out that this feature was discussed in the [sig-arch call on Feb 9th, 2023](https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit#). The implementation was an outcome of that call.

**Review by aravindhp — COMMENTED** [review]
Thanks for the review, @tengqm. I have addressed your comments.

**Comment by aravindhp** [comment]
> @aravindhp could you respond to @tengqm's comments when you have a moment? Tomorrow is the **Docs complete — All PRs reviewed and ready to merge** deadline and we would love to see this merged by then.

@mickeyboxell I have addressed @tengqm's comments.

File: content/en/docs/concepts/cluster-administration/system-logs.md (hunk 11 above)
Comment (aravindhp) [reply content/en/docs/concepts/cluster-administration/system-logs.md:257]: To be clear, `kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/<insert-log-file-name-here>` will work even without this feature.

**Pull request merged**